Lib/test/test_pathlib/support/zip_path.py

```diff
@@ -0,0 +1,245 @@
+import errno
+import io
+import pathlib._abc
+import posixpath
+import stat
+import zipfile
+
+
+class MissingZipPathInfo:
+    """
+    PathInfo implementation that is used when a zip file member is missing.
+    """
+
+    def exists(self, follow_symlinks=True):
+        return False
+
+    def is_dir(self, follow_symlinks=True):
+        return False
+
+    def is_file(self, follow_symlinks=True):
+        return False
+
+    def is_symlink(self, follow_symlinks=True):
+        return False
+
+
+class ZipPathInfo:
+    """
+    PathInfo implementation for an existing zip file member.
+    """
+    __slots__ = ('parent', 'children', 'zip_info')
+
+    def __init__(self, parent=None):
+        self.parent = parent or self
+        self.children = {}
+        self.zip_info = None
+
+    def exists(self, follow_symlinks=True):
+        return True
+
+    def is_dir(self, follow_symlinks=True):
+        if self.zip_info is None:
+            # This directory is implied to exist by another member.
+            return True
+        return self.zip_info.filename.endswith('/')
+
+    def is_file(self, follow_symlinks=True):
+        return not self.is_dir(follow_symlinks=follow_symlinks)
+
+    def is_symlink(self):
+        return False
+
+
+class ZipFileList:
+    """
+    `list`-like object that we inject as `ZipFile.filelist`. We maintain a
+    tree of `ZipPathInfo` objects representing the zip file members. A new
+     `resolve()` method fetches an entry from the tree.
+    """
+
+    __slots__ = ('_root_info', '_items')
+
+    def __init__(self, items):
+        self._root_info = ZipPathInfo()
+        self._items = []
+        for item in items:
+            self.append(item)
+
+    def __len__(self):
+        return len(self._items)
+
+    def __iter__(self):
+        return iter(self._items)
+
+    def append(self, item):
+        self._items.append(item)
+        self.resolve(item.filename, create=True).zip_info = item
+
+    def resolve(self, path, create=False):
+        """
+        Returns a PathInfo object for the given path by walking the tree.
+        """
+        path_info = self._root_info
+        for name in path.split('/'):
+            if not name or name == '.':
+                pass
+            elif name == '..':
+                path_info = path_info.parent
+            elif name in path_info.children:
+                path_info = path_info.children[name]
+            elif create:
+                path_info.children[name] = path_info = ZipPathInfo(path_info)
+            else:
+                path_info = MissingZipPathInfo()
+                break
+        return path_info
+
+
+class ReadableZipPath(pathlib._abc.ReadablePath):
```

Comment:
@jaraco you might find this interesting! For testing purposes only, here's a `zipfile.Path`-like class that's built upon the pathlib ABCs.

It works by replacing `ZipFile.filelist` with a custom `ZipFileList` object that provides its own `append()`. When a `ZipInfo` object is added to the archive, we record it within a tree of `ZipPathInfo` objects. To resolve paths we traverse the tree in `ZipFileList.resolve()`.

There's a nice effect from this: users don't need to provide the canonical zip member path in order to perform queries. e.g. it's possible to do `ZipPath('some_dir', zip_file=...).info.is_dir()` without needing to stick a trailing slash on the end of `some_dir`.